From 10af789a5fb5194ddd5cd31d3ec0ec2634d959e2 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Tue, 28 May 2024 12:15:38 +0200
Subject: [PATCH] irqchip/stm32-exti: Forward ack, set_type and set_wake to
 parent

Propagate missing ack and set_wake to the parent irqchip.

EXTI configurable events are only able to generate a level high
interrupt for the parent interrupt controller (typically the GIC).
In set_type() for configurable events force level high type to
the parent irqchip.

Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I576a6811518a2ebaeb0fb1c6bc123a6b1067cd5c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/383199
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/irqchip/irq-stm32mp-exti.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/irqchip/irq-stm32mp-exti.c b/drivers/irqchip/irq-stm32mp-exti.c
index 0a7ecd4eb684..22f4bb0f76a0 100644
--- a/drivers/irqchip/irq-stm32mp-exti.c
+++ b/drivers/irqchip/irq-stm32mp-exti.c
@@ -399,7 +399,11 @@ static int stm32mp_exti_set_type(struct irq_data *d, unsigned int type)
 unlock:
 	raw_spin_unlock(&chip_data->rlock);
 
-	return err;
+	if (err)
+		return err;
+
+	/* configurable events only propagate level high type to GIC */
+	return irq_chip_set_type_parent(d, IRQ_TYPE_LEVEL_HIGH);
 }
 
 static int stm32mp_exti_set_wake(struct irq_data *d, unsigned int on)
@@ -416,7 +420,7 @@ static int stm32mp_exti_set_wake(struct irq_data *d, unsigned int on)
 
 	raw_spin_unlock(&chip_data->rlock);
 
-	return 0;
+	return irq_chip_set_wake_parent(d, on);
 }
 
 static int stm32mp_exti_suspend(struct device *dev)
@@ -462,6 +466,7 @@ static int stm32mp_exti_retrigger(struct irq_data *d)
 static struct irq_chip stm32mp_exti_chip = {
 	.name			= "stm32mp-exti",
 	.irq_eoi		= stm32mp_exti_eoi,
+	.irq_ack		= irq_chip_ack_parent,
 	.irq_mask		= stm32mp_exti_mask,
 	.irq_unmask		= stm32mp_exti_unmask,
 	.irq_request_resources	= irq_chip_request_resources_parent,
-- 
2.39.5

